### PR TITLE
Hotfix: numercial stability of non-log-stabilized sinkhorn plan

### PR DESCRIPTION
**Pull request by LarsKue**
This is an important hotfix, so sending it directly to main.

### CODE REVIEW

File: bayesflow/utils/optimal_transport/sinkhorn.py [hunk 1]
```diff
@@ -73,7 +73,11 @@ def sinkhorn_plan(
     cost = euclidean(x1, x2)
 
     # initialize the transport plan from a gaussian kernel
-    plan = keras.ops.exp(cost / -(regularization * keras.ops.mean(cost) + 1e-16))
+    logits = cost / -(regularization * keras.ops.mean(cost) + 1e-16)
+    # numerical stability: we can subtract the maximum without changing the result
+    logits = logits - keras.ops.max(logits)
```
[the remark this answers is not in the document]
Comment (LarsKue) [reply bayesflow/utils/optimal_transport/sinkhorn.py]: No.

**Comment by daniel-habermann** [comment]
I have added some additional commits. Most notable changes are:

- `sinkhorn_plan` and `log_sinkhorn_plan` now return proper transport plans, that is, the row and column sums match the marginals.
- the assignments in `sinkhorn` and `log_sinkhorn` are now calculated as `assignments = keras.ops.categorical(log_plan)`, because `keras.ops.categorical` expects log-probs instead of probabilities as inputs (which they confusingly call logits).
- added some additional unit tests to ensure that the transport plans are correct.

@LarsKue could you please double check that the comment in this function definition is what you intended? (the such that.. part)

```python
def sinkhorn(x1: Tensor, x2: Tensor, seed: int = None, **kwargs) -> (Tensor, Tensor):
    """
    Matches elements from x2 onto x1 using the Sinkhorn-Knopp algorithm.

    Sinkhorn-Knopp is an iterative algorithm that repeatedly normalizes the cost matrix into a
    transport plan, containing assignment probabilities.
    The permutation is then sampled randomly according to the transport plan.

    :param x1: Tensor of shape (n, ...)
        Samples from the first distribution.

    :param x2: Tensor of shape (m, ...)
        Samples from the second distribution.

    :param kwargs:
        Additional keyword arguments that are passed to :py:func:`sinkhorn_plan`.

    :param seed: Random seed to use for sampling indices.
        Default: None, which means the seed will be auto-determined for non-compiled contexts.

    :return: Tensor of shape (n,)
        Assignment indices for x2.

    """
    plan = sinkhorn_plan(x1, x2, **kwargs)

    # we sample from log(plan) to receive assignments of length n, corresponding to indices of x2
    # such that x2[assignments] matches x1
    assignments = keras.random.categorical(keras.ops.log(plan), num_samples=1, seed=seed)
    assignments = keras.ops.squeeze(assignments, axis=1)

    return assignments
```

From my point of view this is ready to be merged now.
